packages/neo-one-editor/src/manager/PackageManager.ts

```diff
@@ -67,46 +103,157 @@ export class PackageManager {
+      name,
+      version,
+      packageFiles,
+      subDependencies,
     };
   }
 
-  private async downloadDependency({ name, version, file }: FetchPackageInfo) {
+  private async getAllPackageFiles({ name, version }: FetchPackageInfo): Promise<ReadonlyArray<string>> {
+    const res = await fetch(this.getNPMDataUrl({ name, version }));
+    const fileTree = await res.json();
+
+    return this.getDirectoryList({ dirList: fileTree.files });
+  }
+
+  private getDirectoryList({
+    dirList,
+    currentPath,
+  }: {
+    readonly dirList: ReadonlyArray<FileData | DirectoryData>;
+    readonly currentPath?: string;
+  }): ReadonlyArray<string> {
+    return _.flatten(
+      dirList.map((entry) => {
+        if (entry.type === 'directory') {
+          return this.getDirectoryList({
+            dirList: entry.files,
+            currentPath: currentPath === undefined ? entry.name : `${currentPath}/${entry.name}`,
+          });
+        }
+
+        return currentPath === undefined ? entry.name : `${currentPath}/${entry.name}`;
+      }),
+    );
+  }
+
+  private async checkTypes({ name, version, types }: FetchPackageInfo) {
+    if (types) {
+      return;
+    }
+
+    const typesName = `@types/${name.replace('@', '').replace('/', '__')}`;
+    const versionRes = await fetch(this.getNPMDataUrl({ name: typesName }));
+    if (!versionRes.ok) {
+      return;
+    }
+    const typesVersions = await versionRes.json();
+    const latestTypesVersion = typesVersions.versions[0];
+
+    const res = await fetch(this.getNPMDataUrl({ name: typesName, version: latestTypesVersion }));
+    if (!res.ok) {
+      return;
+    }
+    const files = await res.json();
+    let indexFound = false;
+    files.files.forEach((file: FileData | DirectoryData) => {
+      if (file.name === 'index.d.ts') {
+        indexFound = true;
+      }
+    });
+    if (!indexFound) {
+      return;
+    }
+
+    this.onAddTypes(name, version, `${typesName}@${latestTypesVersion}`);
+  }
+
+  private async startFetchLoop() {
+    if (this.mutableRunning >= this.fetchConcurrency) {
+      return;
+    }
+
+    const entry = this.mutableQueue.shift();
+    // tslint:disable-next-line:no-loop-statement
+    while (entry !== undefined && this.mutableRunning < this.fetchConcurrency) {
+      this.mutableRunning += 1;
+
+      this.downloadDependency({ name: entry.name, version: entry.version, suffix: entry.filePath })
+        .then((file) => {
+          entry.resolve(file);
+          this.mutableRunning -= 1;
+          this.startFetchLoop().catch(() => {
+            // do nothing
+          });
+        })
+        .catch(entry.reject);
+    }
+  }
+
+  private async downloadDependency({ name, version, suffix }: FetchPackageInfo) {
     try {
-      const pkg = await fetch(this.getNPMUrl({ name, version, file }));
+      let minSuffix = suffix;
+      if (suffix !== undefined && suffix.substr(-3, 3) === '.js' && suffix.substr(-7, 4) !== '.min') {
```

Comment:
This is really hard to read.... why not just `suffix.endsWith('.js') && !suffix.endsWith('.min.js')`